README.md
# Easy-push

ES: Te facilita todo el proceso de subir unos cambios a una rama de tu repositorio git,
 añadiendo todos los archivos (git add --all) y mostrando un mensaje donde
 introduciremos el texto del commit.
 <br/>En resumen, hace un push con un commit que escribiremos
 en modo gráfico o por terminal de todos los archivos modificados<br/>
EN: Make a push to your branch, adding all the modified files (git add --all) and shows a message where you can introduce
the text of the commit. In brief, make a push (in master) with a commit that we write in a graphic window.

Ejemplo para ejecutar el script
```shell
gp "Este es el mensaje del commit"
```

## Instalación

La instalación es muy básica, añade dos alias a tu .bashrc

```shell
git clone https://github.com/victormln/easy-push.git
cd easy-push
./install.sh
```

## Modo gráfico/terminal

Modo gráfico:
```shell
gpg "Mensaje del commit"
```

Modo terminal:
```shell
gp "Mensaje del commit"
```

## Configuración

Se pueden configurar varios parámetros. Por ejemplo que no busque actualizaciones automáticas, cambiar el idioma o poner un commit por defecto. Si se abre el archivo **user.conf** se podrán acceder a todas las configuraciones.

## Uso

Este script tiene varios metodos de ejecutarse. Explicaré los comandos para el modo terminal
```shell
gp "Mensaje del commit"
```
OR:
```shell
gp
```

Todos los argumentos disponibles:

|Argumento           |Abreviado|Significado                                   |Uso|
| ------------- | ---- | ---------------------------------------- |----------|
|`"Mensaje del commit"`       |     | Se pondrá este mensaje del commit        |`gp "Mensaje del commit"`  |
|`--conf`     |  | Abre/edita el archivo de configuración del script  |`gp --conf`      |
|`--update`     |  | Busca actualizaciones disponibles.  |`gp --update`      |
|     |`-v`  | Muestra la versión instalada del script.  |`gp -v`      |


changelog.txt
#############
# Changelog #
#############

* Version: 2.1.7 *
#----------------#

- Mejorado instalador

* Version: 2.1.5 *
#----------------#

- Añadido soporte multilenguaje (de momento inglés y español)
- Optimizado update (solo comprobará actualizaciones una vez al dia)
- Al actualizar no se reinicia el script. Mejor que el usuario lo vuelva a ejecutar
- Solucionado bug con el actualizador

* Version: 2.1.2 *
#----------------#

- Solucionados varios bugs
- Comprueba conexión a internet
- Optimizado archivo update.sh
- El commit por defecto que pone el usuario (en el caso que lo ponga), ahora se puede editar al ejecutar el script

* Version: 2.1.1 *
#----------------#

- Solucionado bug con el actualizador
- Comprueba conexión a internet
- Optimizado archivo update.sh

* Version: 2.1.0 *
#----------------#

- Añadidas líneas que muestran la versión del script y el autor al iniciar el script
- Optimizada forma en la que se compara si la versión que se tiene está actualizada
- Añadida opción para que actualice automáticamente a la última versión disponible
- Añadido changelog para ir mostrando las mejoras de las distintas versiones

* Version: 2.0.0 *
#----------------#

- Añadido archivo de configuración user.conf donde el usuario podrá configurar varias configuraciones del script
- Se puede pasar por parámetros el commit que quieres hacer (se pueden poner comillas o no en el mensaje)
- Busca si el script está actualizado y si no es así, da la opción de actualizarlo
- Ahora se puede hacer un push a una rama (antes solo a master)
- Ahora se muestran códigos de colores mostrando en todo momento lo que ha ido bien y lo que ha ido mal
- Se han añadido mensajes de información en todo momento para que no parezca que el script está parado
- Se da la opción de que salga un mensaje emergente cuando acabe de subir los cambios (por defecto desactivado)
- Optimización de código y solucionados varios bugs


terminal/update.sh
#!/bin/bash
# Fitxer: update.sh
# Autor: Víctor Molina Ferreira (victor)
# Data: 26/12/2016
# Versión: 1.0

#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

#  Descripción: Comprueba si el script está a la ultima version

# Última comprobación de una actualización
# (Puede forzar la búsqueda de una actualización con --update)
lastChecked=2017-02-12
if ! command -v curl >/dev/null 2>&1
then
  echo -e "$CURLNOTINSTALLED"
else
  # Compruebo que sistema está usando para hacer ping
  # Si es Linux o Mac
  if [ "$(uname -s)" == "Linux" ] || [ "$(uname)" == "Darwin" ]; then
      ping -c 1 www.google.com &> /dev/null
      has_internet=$(echo $?)
      # Si es Windows
  elif [ "$(expr substr $(uname -s) 1 10)" == "MINGW32_NT" ] ||
    [ "$(expr substr $(uname -s) 1 10)" == "MINGW64_NT" ]; then
      ping -n 1 www.google.com > /dev/null
      has_internet=$(echo $?)
  fi

  today=$(date +%Y-%m-%d)
  if [[ "$today" > "$lastChecked" ]] || [ "$1" == "--update" ]
  then
    # Si el ping se ha realizado correctamente es que tiene internet
    # por lo que se buscaran actualizaciones
    if [ $has_internet -eq 0 ]
    then
      if $search_ota || [ "$1" == "--update" ]
      then
        sed -i 's,^\(lastChecked=\).*,\1'$(date +%Y-%m-%d)',' $( dirname "${BASH_SOURCE[0]}" )/update.sh
        tieneUltimaVersion=false
        # Conseguimos la ultima version que hay en github y le quitamos los puntos
        ultimaVersion=$(curl -s https://raw.githubusercontent.com/victormln/easy-push/master/terminal/autopush.sh | grep '# Versión:' | cut -d: -f 2 | head -1) > /dev/null
        ultimaVersion=${ultimaVersion//[[:blank:]]/}
        ultimaVersionSinPuntos=$( echo $ultimaVersion | tr -d ".")
        # Miramos que versión tiene el usuario actualmente
        versionActualSinPuntos=$(echo $CURRENTVERSION | tr -d ".")
        # Comprobamos si la versionActual es igual o mas grande que la ultimaVersion
        # es igual a la versionActual.
        if (( $versionActualSinPuntos>=$ultimaVersionSinPuntos ))
        then
        	tieneUltimaVersion=true
          if [ "$1" == "--update" ]
          then
            echo -e "$HAVELASTVERSION"
          fi
        else
        	directorioActual=$(pwd)
        	# Nos colocamos en el directorio del script, para actualizarlo
        	cd "$( dirname "${BASH_SOURCE[0]}" )"
        	cd ..
          # Mostramos el mensaje de que hay una nueva actualización
    			echo "###########################################"
    			echo -e "$NEWUPDATEMESSAGE${NC}"
    			echo "$YOUHAVEVERSIONMESSAGE: $CURRENTVERSION"
    			echo "$AVAILABLEVERSIONMESSAGE: $ultimaVersion"
    			echo "###########################################"
        	# Si tiene las actualizaciones automaticas, no se le pide nada
        	if $automatic_update
        	then
            # Si es así, hacemos un pull y le actualizamos el script
    				echo $AVAILABLEVERSIONMESSAGE
            git stash > /dev/null
    				git pull | tee >(echo "$UPDATINGPLEASEWAITMESSAGE")
    				echo -e "$UPDATEDONEMESSAGE"
        	else
            echo "$AVAILABLEUPDATEMESSAGE"
    			  echo "$WANTTODOWNLOADMESSAGE"
        	  # Preguntamos si quiere actualizar
        	  read actualizar
        	  if [ $actualizar == "s" ] || [ $actualizar == "y" ]
        	  then
        			directorioActual=$(pwd)
        			cd "$( dirname "${BASH_SOURCE[0]}" )"
        			cd ..
              git stash > /dev/null
              # Si es así, hacemos un pull y le actualizamos el script
    			  	git pull | tee >(echo "$UPDATINGPLEASEWAITMESSAGE")
    					echo -e "$UPDATEDONEMESSAGE"
              exit
        	  else
              # En el caso que seleccione que no, muestro un mensaje.
    			    echo -e "$NOTUPDATEDMESSAGE"
    					echo -e "**************************"
    					# Damos por su puesto que tiene la ultima version,
    					# para que el script no entre en bucle
    					tieneUltimaVersion=true
        	  fi
        	fi
        	# Cambiamos al directorio donde el usuario tiene sus cambios
        	cd $directorioActual
        fi
      fi
    else
      echo -e "$NOTHAVEINTERNETMESSAGE"
    fi
  fi
fi


terminal/autopush.sh
#!/bin/bash
# Fichero: autopush.sh
# Autor: Víctor Molina Ferreira (victor)
# Fecha: 16/03/16
# Versión: 2.1.7

#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

#  Descripción: Subida de los cambios a un repositorio git (ejecutar el script donde se tenga el .git)

#####################
# Colores de los mensajes
ERROR='\033[0;31m'
OK='\033[0;32m'
WARNING='\033[1;33m'
NC='\033[0m'
# Get the current version
CURRENTVERSION=$(grep '# Versión:' $0 | cut -d: -f 2 | head -1)
CURRENTVERSION=${CURRENTVERSION//[[:blank:]]/}

if [ "$1" == "-v" ]
then
	echo $CURRENTVERSION
	exit
fi
# Cogemos los datos del archivo .conf
source $( dirname "${BASH_SOURCE[0]}" )/user.conf
# Cogemos las variables de idioma
source $( dirname "${BASH_SOURCE[0]}" )/lang/${LANGUAGE}.po

###########
# Body script
####################

echo "Easy Push v$CURRENTVERSION"

if $show_author; then echo "$AUTHORMESSAGE: Víctor Molina [victormln.com] <contact@victormln.com> "; fi;
# Si están activadas las actualizaciones automáticas
# Doy permiso al update.sh
chmod +x $( dirname "${BASH_SOURCE[0]}" )/update.sh

if [ "$1" == "--update" ]
then
	echo -e "$SEARCHINGUPDATESMSG"
elif [ "$1" == "--conf" ]
then
	echo -e "$CONFIGURATIONMSG"
  $default_editor $( dirname "${BASH_SOURCE[0]}" )/user.conf
	exit 0
fi

# Comprobaré si hay alguna versión nueva del programa autopush
# y lo mostraré en pantalla
source $( dirname "${BASH_SOURCE[0]}" )/update.sh

# Comprobamos si existe un .git, si no existe, mostraremos un mensaje y saldremos
if [ ! -e .git ]
then
	echo -e "$NOGITMSG"
	exit 2
fi
# Guardamos en una variable si hay cambios comprobando si hay 3 lineas
# al hacer un git status
hayCambios=$(git status | wc -l)

# En caso de que haya exactamente 3 lineas en el gitstatus,
# significará que no hay nada para añadir o commitear
if [ $hayCambios -eq 3 ]
then
	echo -e "$NOCHANGESMSG"
	exit 3
fi

# Añadimos todos los archivos para subir
git add --all | tee >(echo "$ADDINGFILESMSG")
echo -e "$ALLFILESADDEDMSG"

IFS=''
if [ -z "$commit" ]
then
  # Si el usuario nos ha pasado por parametros un mensaje, lo pondremos como commit
  if [ $# -ne 0 ]
  then
    commit=""
    for var in "$@"
    do
        commit="$commit $var"
    done
  # Sino, le pediremos que mensaje quiere de commit
  else
    # Mostramos un mensaje interactivo donde introduciremos el mensaje
    # a hacer commit
    echo "$INPUTCOMMITMESSAGE"
    read commit
    if [ $? -ne 0 ]
    then
    	exit 1
    fi
  fi
else
	if $allow_edit_default_commit
	then
		read -e -i $commit commit
	fi
fi
# Realizamos el commit con el mensaje puesto en el campo de texto anterior
git commit -m "$commit"

# Guardamos el nombre de la branch actual, para luego hacer un push
# con ese nombre
branch=$(git rev-parse --abbrev-ref HEAD)
# Hacemos un push a origin
git push -u origin $branch

# Mostramos un mensaje conforme ha ido todo bien
# (aquí podria comprobar con un status si es cierto)
if [ $? -eq 0 ]
then
  echo -e "$UPLOADCHANGESDONEMSG"
  # Si tiene activadas las notificaciones, mostraremos un mensaje de Ok
  if $popup_push
  then
    # Mostrabamos un mensaje si los cambios se habian subido bien
    # Añadimos una notificación si se han realizado las subidas al repositorio correctamente
  	zenity --notification --window-icon="info" --text="$CHANGESUPLOADEDMSG" &> /dev/null &
  fi
fi
